tcp.c

```diff
@@ -303,122 +281,149 @@ RD_BOOL
 tcp_tls_connect(void)
 {
 	int err;
-	long options;
 
+	int type;
+	int status;
+	gnutls_datum_t out;
+	gnutls_certificate_credentials_t xcred;
+
+	/* Initialize TLS session */
 	if (!g_ssl_initialized)
 	{
-		SSL_load_error_strings();
-		SSL_library_init();
+		gnutls_global_init();
+		CHECK(gnutls_init(&g_tls_session, GNUTLS_CLIENT));
 		g_ssl_initialized = True;
 	}
 
-	/* create process context */
-	if (g_ssl_ctx == NULL)
-	{
-		g_ssl_ctx = SSL_CTX_new(TLSv1_client_method());
-		if (g_ssl_ctx == NULL)
-		{
-			logger(Core, Error,
-			       "tcp_tls_connect(), SSL_CTX_new() failed to create TLS v1.0 context\n");
-			goto fail;
+	/* It is recommended to use the default priorities */
+	CHECK(gnutls_set_default_priority(g_tls_session));
+	CHECK(gnutls_certificate_allocate_credentials(&xcred));
+	CHECK(gnutls_credentials_set(g_tls_session, GNUTLS_CRD_CERTIFICATE, xcred));
+
+	gnutls_transport_set_int(g_tls_session, g_sock);
+	gnutls_handshake_set_timeout(g_tls_session, GNUTLS_DEFAULT_HANDSHAKE_TIMEOUT);
+
+	/* Perform the TLS handshake */
+	do {
+		err = gnutls_handshake(g_tls_session);
+	} while (err < 0 && gnutls_error_is_fatal(err) == 0);
+
+	if (err < 0) {
+		if (err == GNUTLS_E_CERTIFICATE_VERIFICATION_ERROR) {
+			/* check certificate verification status */
+			type = gnutls_certificate_type_get(g_tls_session);
+			status = gnutls_session_get_verify_cert_status(g_tls_session);
+			CHECK(gnutls_certificate_verification_status_print(status, type, &out, 0));
+			gnutls_free(out.data);
 		}
 
-		options = 0;
-#ifdef SSL_OP_NO_COMPRESSION
-		options |= SSL_OP_NO_COMPRESSION;
-#endif // __SSL_OP_NO_COMPRESSION
-		options |= SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS;
-		SSL_CTX_set_options(g_ssl_ctx, options);
-	}
-
-	/* free old connection */
-	if (g_ssl)
-		SSL_free(g_ssl);
-
-	/* create new ssl connection */
-	g_ssl = SSL_new(g_ssl_ctx);
-	if (g_ssl == NULL)
-	{
-		logger(Core, Error, "tcp_tls_connect(), SSL_new() failed");
 		goto fail;
-	}
 
-	if (SSL_set_fd(g_ssl, g_sock) < 1)
-	{
-		logger(Core, Error, "tcp_tls_connect(), SSL_set_fd() failed");
-		goto fail;
+	} else {
+		char *desc;
+		desc = gnutls_session_get_desc(g_tls_session);
+		//printf("TLS  Session info: %s\n", desc);
```

Comment:
I left it intentionally to show future me or anyone reading this code how to get session info.
Sure It could be ditched.